src/ambianic/pipeline/ai/fall_detect.py

```diff
@@ -243,11 +243,18 @@ def fall_detect(self, image=None):
                 pose_dix = {}
                 inference_result = []
 
-                for e_corr in self.fall_detect_corr:
-                    if pose.keypoints[e_corr].score > 0.6:
-                        # detected key-point probability must be greater than 0.6
-                        pose_dix[e_corr] = pose.keypoints[e_corr].yx
-                                        
+                # Find leftVectorScore and rightVectorScore --> i.e average score of shoulder & hip
+                leftVectorScore = (pose.keypoints['left shoulder'].score + pose.keypoints['left hip'].score) / 2.0
```

Comment:
@bhavikapanara I apologize but I just realized that the average of the hip and shoulder scores is not optimal. What if the hip or shoulder are out of the picture? I've seen users show me images where the model detects a shoulder point correctly, but the hip is some random pixel that is not at all on the body. 

I think we need to apply the confidence threshold to each of the individual hip/shoulder key points. If both hip and shoulder on one side pass the confidence threshold, then we can use the vector, otherwise ignore it.

Making sense?

Reply:
I am attaching a few images that you can use in unit tests to ensure that confidence threshold is used in a meaningful way. I checked the original posenet model demo code (JavaScript) and noticed that it uses two separate parameters: one for the minimal score required for each individual keypoint and another score for the overall average across all keypoints. I think for our case we may be able to get away with applying the confidence threshold (score) to each individual keypoint that represents a hip or shoulder. Ignore other keypoints. Once its implemented we will do more testing with users to verify if that's a stable predictor.

Reply:
![Screen Shot 2021-01-12 at 2 10 23 PM](https://user-images.githubusercontent.com/2234901/104367124-faf91480-54df-11eb-8372-fe4a353db42c.png)
![Screen Shot 2021-01-12 at 2 10 10 PM](https://user-images.githubusercontent.com/2234901/104367128-fb91ab00-54df-11eb-909a-444421629cc9.png)

Reply:
@ivelin  Can you please share only original person image, not screenshot?
I need to feed similar kind of images as we have already used for other tests like
![fall_img_1](https://user-images.githubusercontent.com/19815572/104441763-1fc1ba80-55ba-11eb-865b-72b46b498d30.png)

Reply:
> @bhavikapanara I apologize but I just realized that the average of the hip and shoulder scores is not optimal. What if the hip or shoulder are out of the picture? I've seen users show me images where the model detects a shoulder point correctly, but the hip is some random pixel that is not at all on the body.
> 
> I think we need to apply the confidence threshold to each of the individual hip/shoulder key points. If both hip and shoulder on one side pass the confidence threshold, then we can use the vector, otherwise ignore it.
> 
> Making sense?

So, in that case, our [previous logic](https://github.com/ambianic/ambianic-edge/blob/c418f4db5e6487aa086579ade56347e35ad12c83/src/ambianic/pipeline/ai/fall_detect.py#L247) is correct to check the score individually for hip and shoulder.

One solution can be to consider only hip and shoulder points while calculating the pose_score as we are more interested only in 4 points left and right shoulders and hips instead of t[aking the average of all 17 key-points.](https://github.com/ambianic/ambianic-edge/blob/c418f4db5e6487aa086579ade56347e35ad12c83/src/ambianic/pipeline/ai/pose_engine.py#L166)

For example:
Proposed Solution

```
fall_detect_corr = ['left shoulder', 'left hip', 'right shoulder', 'right hip']

for point_i in range(kps.shape[0]):
    x, y = kps[point_i, 1], kps[point_i, 0]
    prob = self.sigmoid(kps[point_i, 3])

    if KEYPOINTS[point_i] in fall_detect_corr and  prob > 0.60:
        cnt += 1

    keypoint = Keypoint(KEYPOINTS[point_i], [x, y], prob)            
    keypoint_dict[KEYPOINTS[point_i]] = keypoint
    # draw on image and save it for debugging
    draw = ImageDraw.Draw(template_image)
    draw.line(((0,0), (x, y)), fill='red')

pose_scores = cnt/4

```
Instead of 

```
for point_i in range(kps.shape[0]):
    x, y = kps[point_i, 1], kps[point_i, 0]
    prob = self.sigmoid(kps[point_i, 3])

    if prob > 0.60:
        cnt += 1
    keypoint = Keypoint(KEYPOINTS[point_i], [x, y], prob)            
    keypoint_dict[KEYPOINTS[point_i]] = keypoint
    # draw on image and save it for debugging
    draw = ImageDraw.Draw(template_image)
    draw.line(((0,0), (x, y)), fill='red')

pose_scores = cnt/17
```

Reply:
@ivelin any thoughts for the above comment?

Reply:
> @ivelin any thoughts for the above comment?

Here we go:

> So, in that case, our previous logic is correct to check the score individually for hip and shoulder.

Yes, in the sense that we need to apply confidence_threshold to individual keypoints to avoid detections in completely irrelevant locations.

> One solution can be to consider only hip and shoulder points while calculating the pose_score as we are more interested only in 4 points left and right shoulders and hips instead of taking the average of all 17 key-points.

Yes, that is exactly what I am suggesting as a next step.

> Proposed Solution

Yes, that is fine. I actually don't think `pose_score` is relevant in our current fall detection score. 

Instead we have a criteria which requires that we only use a body vector (left or right) to calculate a leaning angle if and only if its individual keypoints (both hip and shoulder) exceed the confidence_threshold (instead of hardcoded 0.6).

In other words `pose_scores` is not used as a limiting criteria for fall detection. 

Maybe a strict definition of `fall_score` can be:

```
# current_body_vector is the one we use for the current angle calculation, either left or right
current_body_vector_score = (hip_score + shoulder_score)/2
leaning_probability = 1 if leaning_angle > fall_leaning_threshold else 0
fall_score = leaning_probability * (previous_body_vector_score + current_body_vector_score)
```

As we've discussed elsewhere, `leaning_probabiliy` is a trivial binary heuristic value for now with `fall_leaning_threshold = 60'. In the future we will likely replace the heuristic rule with a DNN model that is trained to recognize which differences in pose keypoints between image frames constitutes a fall.

Thoughts?

Reply:
@ivelin it makes sense

Reply:
@ivelin I think we also need to update [this part](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L129) as it relies on pose score which is average of all 17 key-points.

It might possible in given image shoulder and the hip point is correctly detected but due to low pose score the image gets rotated and find key-points in the rotated image even in original image hip and shoulder are found.

What's your thoughts?

Reply:
> current_body_vector is the one we use for the current angle calculation, either left or right
> leaning_probability = 1 if leaning_angle > fall_leaning_threshold else 0
> fall_score = leaning_probability * (previous_body_vector_score + current_body_vector_score) / 2

@ivelin Do I need to add this part in current PR?

Reply:
Yes. We need a reasonable value for fall detection score, that will be stored on disk and shown in the UI to the user. Think about the inference output (score, box, metadata) that is expected in store.py from inference classes.

Reply:
> @ivelin I think we also need to update [this part](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L129) as it relies on pose score which is average of all 17 key-points.
> 
> It might possible in given image shoulder and the hip point is correctly detected but due to low pose score the image gets rotated and find key-points in the rotated image even in original image hip and shoulder are found.
> 
> What's your thoughts?

@bhavikapanara  Very good catch.

Your earlier suggestion comes back to light. pose_score is still relevant.

Maybe we can calculate and return pose_score as the average of the left and right vector scores:

```
pose_score = (leftVectorScore + rightVectorScore) / 2
```

This leads me to another concern. What if we see only the lift vector in the previous image and we only see the right vector in the current image? That is still useful information, but in the current logic we would just throw it away and not register any angle change (as both left and right angles will be calculated as non-existant or 0').

Ideally we would work with a learned spinal vector estimate model, but that's a bigger conversation for developing such model or using one that gets us there. See [this recent work](https://ai.googleblog.com/2021/01/recognizing-pose-similarity-in-images.html) from Google AI that may be relevant.


What do you think about using a simple heuristic for now to estimate the spinal vector:
```
if left vector is below confidence threshold, ignore it
if right vector is below confidence threshold, ignore it
if either left or right are above confidence threshold:
  spinalVectorEstimate = average of left and right vectors
if spinalVectorEstimate available:
  calculate leaning angle based on the difference in previous and current spinalVector estimates.
  angle_heuristic_score = 0 or 1 as discussed elsewhere in this PR
  fall_detection_score = angle_heuristic_score * spinalVectorEstimateScore
else:
  no detection
  fall_detection_score = 0
```

@muka I would appreciate your take on this one ^^^

Reply:
@ivelin At this moment the inference output is this: 
[`inference_result.append(('FALL', pose.score, box, angle_change))`](https://github.com/ambianic/ambianic-edge/blob/15d9285ed1ae5f767598a1f2e83dcfca8a43ed50/src/ambianic/pipeline/ai/fall_detect.py#L270)

Now, we want to update it to 
```
leaning_probability = 1 if leaning_angle > fall_leaning_threshold else 0
fall_score = leaning_probability * (previous_body_vector_score + current_body_vector_score) / 2

inference_result.append(('FALL', fall_score, box, ?))

```
Here 

- leaning_angle is [angle_changes](https://github.com/ambianic/ambianic-edge/blob/15d9285ed1ae5f767598a1f2e83dcfca8a43ed50/src/ambianic/pipeline/ai/fall_detect.py#L266) b/w shoulder-hip lines in a current and previous frame
- fall_leaning_threshold --> is [`self._fall_factor` ](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L41)i.e 60 degree 
- previous_body_vector_score: How to find it from [leftVectorScore](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L247) and [rightVectorScore](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L248)
- Similar current_body_vector_score


Also, didn't get you for metadata?

Reply:
> @ivelin At this moment the inference output is this:
> [`inference_result.append(('FALL', pose.score, box, angle_change))`](https://github.com/ambianic/ambianic-edge/blob/15d9285ed1ae5f767598a1f2e83dcfca8a43ed50/src/ambianic/pipeline/ai/fall_detect.py#L270)
> 
> Now, we want to update it to
> 
> ```
> leaning_probability = 1 if leaning_angle > fall_leaning_threshold else 0
> fall_score = leaning_probability * (previous_body_vector_score + current_body_vector_score) / 2
> 
> inference_result.append(('FALL', fall_score, box, ?))
> ```

Yes.

> 
> Here
> 
> * leaning_angle is [angle_changes](https://github.com/ambianic/ambianic-edge/blob/15d9285ed1ae5f767598a1f2e83dcfca8a43ed50/src/ambianic/pipeline/ai/fall_detect.py#L266) b/w shoulder-hip lines in a current and previous frame

Yes.

> * fall_leaning_threshold --> is [`self._fall_factor` ](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L41)i.e 60 degree

Yes.

> * previous_body_vector_score: How to find it from [leftVectorScore](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L247) and [rightVectorScore](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L248)

Here I meant the body vector score calculated for the previous image and the body vector score calculated for the current image.

> * Similar current_body_vector_score

This now became a new topic after you discovered the issue in the rotation logic. Let me know your thoughts on spine estimation and pose_score formulas.

> 
> Also, didn't get you for metadata?

That's already taken care of [here](https://github.com/ambianic/ambianic-edge/blob/15d9285ed1ae5f767598a1f2e83dcfca8a43ed50/src/ambianic/pipeline/ai/fall_detect.py#L65). It's just decorative data to display in the UI for user convenience.

Reply:
Just realized that `some_angle_correction` is not trivial, because we don't know if the image is from the front, back or side of the person since its only one side vector and we have no simple reference point. Let's skip this angle correction for now. And in the near future look closer at transfer learning a DNN model such as [Pr_VIPE](https://ai.googleblog.com/2021/01/recognizing-pose-similarity-in-images.html) for 3D pose retrieval from monocular pose 2D images.

Reply:
> Here I meant the body vector score calculated for the previous image and the body vector score calculated for the current image.


@ivelin Still I didn't get which body vector you are referencing here!
how to calculate **previous_body_vector_score** and **current_body_vector_score** for previous and current image? Can you please elaborate it with mathematical representation?

Reply:
@bhavikapanara 

The latest iteration is as follows:

# see formula for spinalVectorScore above
current_body_vector_score  = pose_score = spinalVectorScore
# do angle calculation
# do fall_score calculation

# save current_body_vector_score along with keypoints as previous and compare when the next image frame comes in
previous_body_vector_score = current_body_vector_score

Reply:
@ivelin 
Here required to change the condition to manage the image [rotation section](https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L129) by comparing newly invented pose_score instead of pose.score

https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L129

will become

        while (not poses or pose_score < min_score) and rotations:

And

https://github.com/ambianic/ambianic-edge/blob/279905c058f22337e3fe05e9e6f3b3bc559f5648/src/ambianic/pipeline/ai/fall_detect.py#L137
will become

        if (pose and pose_score >= min_score):

right?